src/lib/consensus/proof_of_stake.ml

```diff
@@ -2572,30 +2607,36 @@ module Hooks = struct
           (Coda_base.Sparse_ledger.merkle_root snapshot.ledger) ;
         snapshot
       in
-      let proposal_data slot =
-        (* Try vrfs for all keypairs within this slot until one wins or all lose *)
-        Keypair.Set.fold_until keypairs ~init:()
-          ~f:(fun () keypair ->
-            match
-              Vrf.check ~epoch ~slot ~seed:epoch_data.seed ~epoch_snapshot
-                ~private_key:keypair.private_key ~public_key:keypair.public_key
-                ~total_stake ~logger
-            with
-            | None ->
-                Continue_or_stop.Continue ()
-            | Some data ->
-                Continue_or_stop.Stop (Some (keypair, data)) )
+      let proposal_data unseen_pks slot =
+        (* Try vrfs for all keypairs that are unseen within this slot until one wins or all lose *)
+        (* TODO: Don't do this, and instead pick the one that has the highest
+         * chance of winning *)
+        Keypair.And_compressed_pk.Set.fold_until keypairs ~init:()
+          ~f:(fun () (keypair, public_key_compressed) ->
+            if Public_key.Compressed.Set.mem unseen_pks public_key_compressed
+            then Continue_or_stop.Continue ()
+            else
+              match
+                Vrf.check ~epoch ~slot ~seed:epoch_data.seed ~epoch_snapshot
+                  ~private_key:keypair.private_key
+                  ~public_key:keypair.public_key ~public_key_compressed
+                  ~total_stake ~logger
+              with
+              | None ->
+                  Continue_or_stop.Continue ()
+              | Some data ->
+                  Continue_or_stop.Stop (Some (keypair, data)) )
           ~finish:(fun () -> None)
       in
       let rec find_winning_slot (slot : Epoch.Slot.t) =
         if UInt32.of_int (Epoch.Slot.to_int slot) >= Constants.Epoch.size then
           None
         else
           match Local_state.seen_slot local_state epoch slot with
-          | `Seen ->
+          | `All_seen ->
               find_winning_slot (Epoch.Slot.succ slot)
-          | `Unseen -> (
-            match proposal_data slot with
+          | `Unseen pks -> (
```

Comment:
You still need to recursively invoke `find_winning_slot` for the pks that have seen all the slots.

Reply:
As discussed, this should work